feat(avm)!: shifts skeleton

IlyasRidhuan opened the pull request:
Skeleton for shift opcodes to get further in the bulk test..

fyi @MirandaWood 

On Mainframe
```
## TX Label: bulk_test/0
- Total duration: `287.005 ms`
- Total mana used: `441,601`
- Mana per second: `1,538,654`
- Total instructions executed: `25,224`
- Private insertions:
    - Non-revertible: `7.847 ms`
    - Revertible: `7.047 ms`
- Proving:
    - Simulation (all): `241 ms`
    - Proving (all): `20,071 ms`
    - Trace generation (all): `3,032 ms`
    - Trace generation interactions: `1,515 ms`
    - Trace generation traces: `1,505 ms`
    - Sumcheck: `5,488 ms`
    - PCS: `3,414 ms`
    - Log derivative inverse: `369 ms`
    - Log derivative inverse commitments: `2,847 ms`
    - Wire commitments: `6,245 ms`
- Enqueued public calls:
    - **Fn: AvmTest.bulk_testing**
        - Duration: `246.462 ms`
        - Mana used: `441,601`
        - Mana per second: `1,791,764`
        - Instructions executed: `25,224`

IlyasRidhuan commented:
* **#16296** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16296?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16295** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16295?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16295?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#16164** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16164?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16151** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16151?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

IlyasRidhuan commented on barretenberg/cpp/src/barretenberg/vm2/simulation/alu.cpp:
```diff
@@ -187,4 +187,30 @@ MemoryValue Alu::truncate(const FF& a, MemoryTag dst_tag)
+{
+    // todo: Tag checks need to change to ensure LHS == RHS, this is in the ShiftVisitor in tagged value
+    try {
+        MemoryValue c = a << b;
```
This only works while we expect RHS to be U8

fcarreiro replied:
I'm surprised this works. In the version I sent you I had
```
 MemoryValue c = a >> MemoryValue::from_tag(a_tag, b.as<FF>());
```
What does `a << b` do wrt types if you have say u32 and u8?

IlyasRidhuan replied:
I think the `ShiftOperationVisitor` is designed to allow the RHS to be different to LHS. When the noir changes land in next we can get rid of `ShiftOperationVisitor` and go back to `BinaryOperationVisitor`.

fcarreiro replied:
That seems to be right... Though the tests (`ShiftOperationsWithDifferentTypes`) are wrong because I (well Cursor) changed the RHS and not LHS. Just to be sure could you change the LHS and assert that the final type is the RHS one?

IlyasRidhuan replied:
>assert that the final type is the RHS one?

i assume that you mean LHS here?

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/simulation/execution.cpp:
```diff
@@ -955,6 +955,44 @@ void Execution::sha256_compression(ContextInterface& context,
+    MemoryValue b = memory.get(b_addr);
+    set_and_validate_inputs(opcode, { a, b });
+
+    get_gas_tracker().consume_gas();
```
Isn't gas dynamic? I sent you a snippet with (what I hoped) was a full impl

```
uint8_t get_bitwise_dynamic_gas_multiplier(MemoryTag tag)
{
    switch (tag) {
    case MemoryTag::U8:
        return 1;
    case MemoryTag::U16:
        return 2;
    case MemoryTag::U32:
        return 4;
    case MemoryTag::U64:
        return 8;
    case MemoryTag::U128:
        return 16;
    case MemoryTag::U1:
        return 1;
    case MemoryTag::FF:
        return 0;
    default:
        throw OpcodeExecutionException("Invalid tag type for gas cost multiplier: " +
                                       std::to_string(static_cast<uint8_t>(tag)));
    }
}

// TODO(miranda): constrain and do the right thing with rhs param.
void Execution::shl(ContextInterface& context, MemoryAddress a_addr, MemoryAddress b_addr, MemoryAddress dst_addr)
{
    constexpr auto opcode = ExecutionOpCode::SHL;
    auto& memory = context.get_memory();
    MemoryValue a = memory.get(a_addr);
    MemoryValue b = memory.get(b_addr);
    set_and_validate_inputs(opcode, { a, b });

    get_gas_tracker().consume_gas({ .l2Gas = get_bitwise_dynamic_gas_multiplier(a.get_tag()), .daGas = 0 });

    // Only allow integer tags for shift operations
    if (a.get_tag() == MemoryTag::FF || b.get_tag() == MemoryTag::FF) {
        throw OpcodeExecutionException("SHL: Shift not valid for FF tag");
    }
    // Use the underlying value types for the shift
    auto a_tag = a.get_tag();
    MemoryValue c = a << MemoryValue::from_tag(a_tag, b.as<FF>());
    memory.set(dst_addr, c);
    set_output(opcode, c);
}

// TODO(miranda): constrain and do the right thing with rhs param.
void Execution::shr(ContextInterface& context, MemoryAddress a_addr, MemoryAddress b_addr, MemoryAddress dst_addr)
{
    constexpr auto opcode = ExecutionOpCode::SHR;
    auto& memory = context.get_memory();
    MemoryValue a = memory.get(a_addr);
    MemoryValue b = memory.get(b_addr);
    set_and_validate_inputs(opcode, { a, b });

    get_gas_tracker().consume_gas({ .l2Gas = get_bitwise_dynamic_gas_multiplier(a.get_tag()), .daGas = 0 });

    // Only allow integer tags for shift operations
    if (a.get_tag() == MemoryTag::FF || b.get_tag() == MemoryTag::FF) {
        throw OpcodeExecutionException("SHR: Shift not valid for FF tag");
    }
    // Use the underlying value types for the shift
    auto a_tag = a.get_tag();
    MemoryValue c = a >> MemoryValue::from_tag(a_tag, b.as<FF>());
    memory.set(dst_addr, c);
    set_output(opcode, c);
}
```

sirasistant replied:
It's true that if we don't include dynamic gas in the skeleton then end gas won't match in the bulk test for sure
@IlyasRidhuan if you add this remember to update the instruction spec with the dyn gas constants

IlyasRidhuan replied:
wait is shift dynamically costed based on the bitwise dynamic gas? Sorry i saw bitwise amd assumed this was a cursor thing haha

IlyasRidhuan commented on yarn-project/simulator/src/public/avm/opcodes/bitwise.ts:
```diff
@@ -79,32 +79,6 @@ export class Xor extends ThreeOperandBitwiseInstruction {
   }
 }
 
-export class Shl extends ThreeOperandBitwiseInstruction {
```
Moved to `arithmetic.ts`, this removes the dynamic gas component

IlyasRidhuan commented on barretenberg/cpp/src/barretenberg/vm2/common/tagged_value.test.cpp:
```diff
@@ -508,25 +508,25 @@ TEST(TaggedValueTest, ErrorCases)
-    EXPECT_EQ(result_shl_u1.as<uint32_t>(), 2);
+    auto result_shl_u1 = u1_amount << shift_val;
+    ASSERT_EQ(result_shl_u1.get_tag(), ValueTag::U1);
+    EXPECT_EQ(static_cast<uint8_t>(result_shl_u1.as<uint1_t>()), 0); // 1 << 3 = 0 with overflow
```
These test have been changed so the RHS is kept as u8 (for now) and the LHS is changed. Tag check also changes to `ASSERT_EQ` becuase if it fails, then the subsequent `EXPECT_EQ` that utilises `.as()` will fail.

btw @fcarreiro had to static_cast the `.as<uint1_t>()` since couldn't get the equality to compile ( does `.as()` return a bool for U1?)

fcarreiro replied:
hmm I thought uint1 is it's own type now, maybe that's why? what if you do `static_cast<uint1_t>(0)` instead of static casting the lhs?

IlyasRidhuan committed:
feat(avm)!: shifts skeleton

Skeleton for shift opcodes to get further in the bulk test..

fyi @MirandaWood

On Mainframe
```
## TX Label: bulk_test/0
- Total duration: `287.005 ms`
- Total mana used: `441,601`
- Mana per second: `1,538,654`
- Total instructions executed: `25,224`
- Private insertions:
    - Non-revertible: `7.847 ms`
    - Revertible: `7.047 ms`
- Proving:
    - Simulation (all): `241 ms`
    - Proving (all): `20,071 ms`
    - Trace generation (all): `3,032 ms`
    - Trace generation interactions: `1,515 ms`
    - Trace generation traces: `1,505 ms`
    - Sumcheck: `5,488 ms`
    - PCS: `3,414 ms`
    - Log derivative inverse: `369 ms`
    - Log derivative inverse commitments: `2,847 ms`
    - Wire commitments: `6,245 ms`
- Enqueued public calls:
    - **Fn: AvmTest.bulk_testing**
        - Duration: `246.462 ms`
        - Mana used: `441,601`
        - Mana per second: `1,791,764`
        - Instructions executed: `25,224`
